### DISCUSSION TITLE
How to reset signal when re-rendering of a child component of the same type in the same location?

### DISCUSSION THREAD

**Discussion by ixmoyren** [DioxusLabs/dioxus#2864] [Q&A] [answered]
In dioxus version 0.6.0-alpha-2, a re-rendering was triggered by a change in the Signal of the parent component, but the re-rendering of a child component of the same type in the same location only updated the value corresponding to the incoming Props, but did not reset the Signal within the child component. How can I make the Signal in the component to be reset when re-rendering?

**Comment by ixmoyren** [comment]
I tried to add a key attribute to the component, like [react does](https://react.dev/learn/preserving-and-resetting-state#option-2-resetting-state-with-a-key), but it doesn't work in dioxus!

**Comment by ixmoyren** [comment, marked as answer]
Currently, dixous' key only works in lists. To solve the above problem, you can build the component in an iterator, like [this]( https://github.com/DioxusLabs/docsite/blob/ffafa9e366616ed77a5691fa7bae515a92933fdf/src/doc_examples/component_lifecycle.rs#L37-L41 )

**Comment by ixmoyren** [comment]
After this [pr](https://github.com/DioxusLabs/dioxus/pull/2886 ), key-prop works in normal components without the need to use it in iterators.

**Reply by ealmloff** [reply]
> After this [pr](https://github.com/DioxusLabs/dioxus/pull/2886), key-prop works in normal components without the need to use it in iterators.

Keys still only matter in lists. The PR you linked just fixes hot reloading keys in debug mode

**Reply by ixmoyren** [reply]
en... I'm a little confused about the definition of this debug mode. I use the `dx build` command to build and then start a simple server to run. At this point, key-prop(https://github.com/ixmoyren/dioxus-example/blob/main/src/main.rs#L99) is enabling re-rendering.

The repo

https://github.com/ixmoyren/dioxus-example/tree/main

 The operation
```bash
# under the project directory
dx build
cd dist
python -m http.server 9000 --bind 127.0.0.1
```

**Reply by ixmoyren** [reply]
FormSplitBill(https://github.com/ixmoyren/dioxus-example/blob/main/src/main.rs#L99) with key: 
[Screencast_20240907_011548.webm](https://github.com/user-attachments/assets/d532d921-f163-436f-a1c1-65e9cd1d5c13)

FormSplitBill without key:
[Screencast_20240907_012317.webm](https://github.com/user-attachments/assets/0b0e2b2d-159f-4db1-99e2-049e4c144a57)

**Reply by ealmloff** [reply]
I just mean building without the `--release` flag. The key attribute got broken in some hot reloading upgrades which are only compiled in debug mode.

The key definitely fixes things. IIRC if statements were also changed from desugering into a placeholder or the value to desugering into an iterator. That might have been the changed that fixed this.

**Reply by ixmoyren** [reply]
In this example, the output built by `dx build` or `dx build --release` has the same runtime effect. I used another similar example to verify the key attribute. The key attribute allows the component in the example to re-render, likely due to the if statement.
